### PR TITLE
Migrate subscription operations to listener module

### PR DESCRIPTION
**Pull request by Chuhaa**
## Purpose

Migrate subscription operations to listener module from client module.

## Goal
 Fix https://github.com/wso2-enterprise/choreo/issues/4777

### CODE REVIEW

File: calendar/modules/listener/watch_scheduler.bal [hunk 1]
```diff
@@ -1,17 +1,59 @@
...
+        error? err = self.scheduleNextChannelRenewal();
+        if (err is error) {
+            if (self.retryScheduleCount < 10) {
+                runtime:sleep(5);
+                self.retryScheduleCount += 1;
+                self.scheduleNextChannel();
+            } else {
+                panic error(ERR_SCHEDULE);
+            }
+        }
```
Comment (Chuhaa) [inline calendar/modules/listener/watch_scheduler.bal:56]: @abeykoon @SkNuwanTissera Could you please review this?

File: calendar/modules/listener/constants.bal [hunk 2]
```diff
@@ -43,4 +43,6 @@ const string HTTP_ERROR_MSG = "Error occurred while getting the HTTP response :
 const string ERR_EXTRACTING_ERROR_MSG = "Error occured while extracting errors from payload.";
 const string JSON_ACCESSING_ERROR_MSG = "Error occurred while accessing the JSON payload of the response.";
 const string ERR_WATCH_RESPONSE = "Error occurred while constructing WatchResponse record.";
+const string ERR_CHANNEL_REGISTRATION = "Unable to register new channel.";
+const string ERR_SCHEDULE = "Unable to schedule next subscription.";
```
Comment (abeykoon) [inline calendar/modules/listener/constants.bal:47]: subscription renewal?

File: calendar/modules/listener/watch_scheduler.bal [hunk 3]
```diff
@@ -14,20 +14,46 @@
...
         self.'listener = 'listener;
     }
 
     public isolated function execute() {
-        checkpanic self.'listener.registerWatchChannel();
-        checkpanic self.scheduleNextChannelRenewal();
+        error? err = self.'listener.registerWatchChannel();
+        if (err is error) {
+            if (self.retryCount < 10) {
+                runtime:sleep(5);
```
Comment (abeykoon) [inline calendar/modules/listener/watch_scheduler.bal:35]: 5 means in miliseconds? It is too low if that is the case.

File: calendar/modules/listener/watch_scheduler.bal [hunk 4]
```diff
@@ -14,20 +14,46 @@
...
+                self.retryCount += 1;
+                self.execute();
```
Comment (abeykoon) [inline calendar/modules/listener/watch_scheduler.bal:37]: I guess this is a recursive call?

File: calendar/modules/listener/watch_scheduler.bal [hunk 5]
```diff
@@ -14,20 +14,46 @@
...
+            } else {
+                panic error(ERR_CHANNEL_REGISTRATION);
+            }
+        } else {
+            self.scheduleNextChannel();
+        }
+    }
+
+    isolated function scheduleNextChannel() {
+        error? err = self.scheduleNextChannelRenewal();
```
Comment (abeykoon) [inline calendar/modules/listener/watch_scheduler.bal:47]: What kind of error we can get in here?

File: calendar/modules/listener/watch_scheduler.bal [hunk 6]
```diff
@@ -14,20 +14,46 @@
...
 
     isolated function init(Listener 'listener) {
         self.'listener = 'listener;
     }
 
     public isolated function execute() {
-        checkpanic self.'listener.registerWatchChannel();
-        checkpanic self.scheduleNextChannelRenewal();
+        error? err = self.'listener.registerWatchChannel();
+        if (err is error) {
```
Comment (abeykoon) [inline calendar/modules/listener/watch_scheduler.bal:33]: Shall we print a WARN log saying `Could not register watch channel`. Something like 
`log.warn("Could not register watch channel" +  error.toString());`

File: calendar/modules/listener/watch_scheduler.bal [hunk 7]
```diff
@@ -14,20 +14,46 @@
...
+            if (self.retryCount < 10) {
+                runtime:sleep(5);
+                self.retryCount += 1;
```
Comment (abeykoon) [inline calendar/modules/listener/watch_scheduler.bal:36]: Shall we print a log saying we are retrying to register watch channel - retry count is = x

Comment (Chuhaa) [reply calendar/modules/listener/watch_scheduler.bal:39]: Fixed in bb0c08e9a7f7da0b076f584b770b90e7d7a06aa4

File: calendar/modules/listener/constants.bal (hunk 2 above)
Comment (Chuhaa) [reply calendar/modules/listener/constants.bal]: Fixed in bb0c08e9a7f7da0b076f584b770b90e7d7a06aa4

**Review by abeykoon — APPROVED** [review]

**Pull request merged**